core/Pos/Core/Address.hs

```diff
@@ -346,6 +347,46 @@ deriveLvl2KeyPair ibea scp passphrase wsKey accountIndex addressIndex = do
     -- We don't need to check passphrase twice
     createHDAddressH ibea (ShouldCheckPassphrase False) passphrase hdPass wKey [accountIndex] addressIndex
```

Comment:
This is just a thought, but if I squint my eyes I can see similarities between `createHDAddressH` and your `deriveHDAddressKeyPair`. With a bit of refactoring, in theory, we should be able to generalise `createHDAddressH` to simply create addresses of arbitrary depth, and then our `deriveLvl2KeyPair` and `deriveLvl3KeyPair` might become trivial functions which offload everything to `deriveHDAddressH`.

Do you think I'm being mental here? 😀

Reply:
Totally agree here. I was a little worried to make too many changes to others code but certainly deriveHDAddressKeyPair is meant to be general. I will make the changes such that the functions use the same code.

Reply:
Thanks man! 😉